### ISSUE TITLE
Tuning with Raytune doesn't save checkpoints nor plots

### ISSUE THREAD

**Issue by Dapid** [ultralytics/ultralytics#16317]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Hyperparameter Tuning

### Bug

I used the following test script, minimally modified from the documentation:

```python
from ultralytics import YOLO

RAY = True
# Initialize the YOLO model
model = YOLO("yolov8n.pt")

# Tune hyperparameters on COCO8 for 30 epochs
model.tune(data="coco8.yaml", epochs=3, iterations=3, grace_period=2, optimizer="AdamW", plots=True, save=True,
           project="mycoco", use_ray=RAY)
```

When `RAY` is `False`, everything works as expected, producing plots and checkpoints. On the other hand, when raytune is enabled, I get neither.

```
$ ls
run_coco.py  runs  yolov8n.pt
$ ls runs/detect/tune/_tune_2024-09-17_10-35-28/
 basic-variant-state-2024-09-17_10-35-28.json
 experiment_state-2024-09-17_10-35-28.json
'_tune_cadf0_00000_0_bgr=0.1896,box=0.0950,cls=1.1936,copy_paste=0.3934,degrees=5.0322,fliplr=0.9175,flipud=0.0554,hsv_h=0.0121,hsv_2024-09-17_10-35-28'
'_tune_cadf0_00001_1_bgr=0.7049,box=0.0747,cls=2.3577,copy_paste=0.9653,degrees=37.2006,fliplr=0.2743,flipud=0.2230,hsv_h=0.0979,hs_2024-09-17_10-35-28'
'_tune_cadf0_00002_2_bgr=0.1109,box=0.0940,cls=2.9258,copy_paste=0.6203,degrees=17.6681,fliplr=0.9465,flipud=0.3761,hsv_h=0.0123,hs_2024-09-17_10-35-28'
 tuner.pkl
$ ls runs/detect/tune/_tune_2024-09-17_10-35-28/_tune_cadf0_0000*
'runs/detect/tune/_tune_2024-09-17_10-35-28/_tune_cadf0_00000_0_bgr=0.1896,box=0.0950,cls=1.1936,copy_paste=0.3934,degrees=5.0322,fliplr=0.9175,flipud=0.0554,hsv_h=0.0121,hsv_2024-09-17_10-35-28':
events.out.tfevents.1726562129.Hopper  params.json  params.pkl  progress.csv  result.json

'runs/detect/tune/_tune_2024-09-17_10-35-28/_tune_cadf0_00001_1_bgr=0.7049,box=0.0747,cls=2.3577,copy_paste=0.9653,degrees=37.2006,fliplr=0.2743,flipud=0.2230,hsv_h=0.0979,hs_2024-09-17_10-35-28':
events.out.tfevents.1726562151.Hopper  params.json  params.pkl  progress.csv  result.json

'runs/detect/tune/_tune_2024-09-17_10-35-28/_tune_cadf0_00002_2_bgr=0.1109,box=0.0940,cls=2.9258,copy_paste=0.6203,degrees=17.6681,fliplr=0.9465,flipud=0.3761,hsv_h=0.0123,hs_2024-09-17_10-35-28':
events.out.tfevents.1726562172.Hopper  params.json  params.pkl  progress.csv  result.json

```

There are references to the project folder, such as

```
(_tune pid=1204893) TensorBoard: Start with 'tensorboard --logdir mycoco/train', view at http://localhost:6006/
(_tune pid=1204893) Plotting labels to mycoco/train/labels.jpg
```
but no `mycoco` folder can be found.

### Environment

```
$ yolo checks
Ultralytics YOLOv8.2.95 🚀 Python-3.10.15 torch-2.4.1+cu124 CUDA:0 (NVIDIA GeForce RTX 2070, 7790MiB)
Setup complete ✅ (8 CPUs, 62.7 GB RAM, 704.3/893.2 GB disk)

OS                  Linux-6.10.6-200.fc40.x86_64-x86_64-with-glibc2.39
Environment         Linux
Python              3.10.15
Install             pip
RAM                 62.72 GB
CPU                 Intel Core(TM) i7-9700F 3.00GHz
CUDA                12.4

numpy               ✅ 1.26.4<2.0.0,>=1.23.0
matplotlib          ✅ 3.9.2>=3.3.0
opencv-python       ✅ 4.10.0.84>=4.6.0
pillow              ✅ 10.4.0>=7.1.2
pyyaml              ✅ 6.0.2>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.14.1>=1.4.1
torch               ✅ 2.4.1+cu124>=1.8.0
torchvision         ✅ 0.19.1+cu124>=0.9.0
tqdm                ✅ 4.66.5>=4.64.0
psutil              ✅ 6.0.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.2>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.6>=2.0.0
torch               ✅ 2.4.1+cu124!=2.4.0,>=1.8.0; sys_platform == "win32"
```

### Minimal Reproducible Example

```python
from ultralytics import YOLO

RAY = True
# Initialize the YOLO model
model = YOLO("yolov8n.pt")

# Tune hyperparameters on COCO8 for 30 epochs
model.tune(data="coco8.yaml", epochs=3, iterations=3, grace_period=2, optimizer="AdamW", plots=True, save=True,
           project="mycoco", use_ray=RAY)
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hi @Dapid, thank you for reaching out and for your interest in Ultralytics 🚀!

We recommend new users check out our [Docs](https://docs.ultralytics.com) for various [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples. Many common questions and issues are addressed there.

### Bug Report 🐛

It looks like you're experiencing an issue with Raytune not saving checkpoints or plots. Since you've provided a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/), an Ultralytics engineer will assist you soon.

### Community & Support 💬

While waiting, feel free to join the Ultralytics community to engage in real-time chat on [Discord](https://ultralytics.com/discord) 🎧, explore more in-depth discussions on [Discourse](https://community.ultralytics.com), or connect with peers on our [Subreddit](https://reddit.com/r/ultralytics).

### Upgrade Instructions

In the meantime, please ensure you're using the latest `ultralytics` package with its [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) and [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) environment to check if the issue is already fixed:

```bash
pip install -U ultralytics
```

### Environments

YOLOv8 can be run across various environments, including:

- **Notebooks with free GPU**: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov8"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>

- **Google Cloud**: See our [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)

- **Amazon**: Check the [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)

- **Docker**: Refer to the [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

### Status

If the [CI Badge](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) is green, all tests pass, ensuring correct operation of YOLOv8 on macOS, Windows, and Ubuntu.

Your patience is appreciated, and further assistance will follow soon!

**Comment by Y-T-G** [comment]
By default, Ray tune stores the files in `/tmp/ray/`

**Comment by Dapid** [comment]
I see, but now ultralytics is also hijacking that folder for some of its outputs, but not all, in a way that isn't neither documented nor predictable, plus the information on stdout is wrong.

This is a bad idea, because anything in `/tmp` could disappear, and there is no easy way to get the path to the folders to save the data. If there is a power outage during the training, I will loose all my progress, and won't be able to access the trained models, unless I have a parallel daemon saving those.

This is also a change in behaviour, it used to save the things in the running folder.

To be clear, I don't object to Raytune writing to its temp folder, but to Ultralytics writing data that I explicitly requested there. And it seems it doesn't even know the checkpoints are saved: here is what the result grid knows.

```
ResultGrid<[
  Result(
    metrics={'metrics/precision(B)': 0.7035117314342072, 'metrics/recall(B)': 0.7377450980392157, 'metrics/mAP50(B)': 0.7773131240850301, 'metrics/mAP50-95(B)': 0.3688889952348491, 'fitness': 0.4097314081198672},
    path=[my local path],
    filesystem='local',
    checkpoint=None
  ),
]
```

**Comment by Y-T-G** [comment]
The training is delgated to Raytune and Raytune is running ultralytics training by using the temporary folder as working directory. The default behavior of ultralytics is to create the runs inside working directory where you're running the training from, which is why the runs are saved there.

https://github.com/ray-project/ray/issues/9571

It's not an ultralytics specified behavior, but rather something as a result of the fact that the training was delegated to Raytune.

**Comment by Y-T-G** [comment]
As per their [docs](https://docs.ray.io/en/latest/tune/api/doc/ray.tune.Trainable.html), you can turn off this behavior by setting the environment variable `RAY_CHDIR_TO_TRIAL_DIR=0`

**Comment by glenn-jocher** [comment]
@Y-T-G should we try to set this ENV directly or pass the `project` and `name` args that create the save_dir to Ray?

@Dapid Ray is a separate tool that is running it's own processes outside of our control, but feel free to submit a PR to adapt the behaviour to what you were expecting.

**Comment by Y-T-G** [comment]
@glenn-jocher I tested it with the environment variable set, and it created the training folders in the usual runs directory, albeit it was creating folders called `detect/trainN` in it instead of using the `tune` subfolders. Maybe we could pass an identifiable `project` name for the session so that they get collected inside the same subfolder.

**Comment by Dapid** [comment]
@glenn-jocher I understand that Ray files are stored elsewhere, but plots like the correlogram, or `best.pt`
 
![labels_correlogram](https://github.com/user-attachments/assets/d53a120c-9c1d-4f2d-aefc-64bac9cfd533)

are generated by Ultralytics, and I wouldn't expect them to land on some odd folder, ignoring the `project` variable. And indeed, when I run the same code a few months ago, the ultralytics-generated files landed in the expected folder.

It seems the problem is that Raytune chdirs to its own folder, but then ultralytics is not aware of it. Quoting again from the stdout: 

```
(_tune pid=1204893) TensorBoard: Start with 'tensorboard --logdir mycoco/train', view at http://localhost:6006/
(_tune pid=1204893) Plotting labels to mycoco/train/labels.jpg
```

I wouldn't expect `mycoco/train` to _actually_ mean `/tmp/ray/session_2024-09-17_09-58-00_368067_1187636/artifacts/2024-09-17_09-58-01/_tune_2024-09-17_09-58-01/working_dirs/_tune_9000a_00000_0_bgr=0.4437,box=0.0814,cls=3.5071,copy_paste=0.7000,degrees=37.3318,fliplr=0.7200,flipud=0.5856,hsv_h=0.0168,hs_2024-09-17_09-58-02/mycoco/train`

The solution is, I think, to store internally absolute paths, so we aren't affected by RayTune.

**Comment by Dapid** [comment]
@glenn-jocher I understand that Ray files are stored elsewhere, but plots like the correlogram, or `best.pt`
 
![labels_correlogram](https://github.com/user-attachments/assets/d53a120c-9c1d-4f2d-aefc-64bac9cfd533)

are generated by Ultralytics, and I wouldn't expect them to land on some odd folder, ignoring the `project` variable. And indeed, when I run the same code a few months ago, the ultralytics-generated files landed in the expected folder.

It seems the problem is that Raytune chdirs to its own folder, but then ultralytics is not aware of it. Quoting again from the stdout: 

```
(_tune pid=1204893) TensorBoard: Start with 'tensorboard --logdir mycoco/train', view at http://localhost:6006/
(_tune pid=1204893) Plotting labels to mycoco/train/labels.jpg
```

I wouldn't expect `mycoco/train` to _actually_ mean `/tmp/ray/session_2024-09-17_09-58-00_368067_1187636/artifacts/2024-09-17_09-58-01/_tune_2024-09-17_09-58-01/working_dirs/_tune_9000a_00000_0_bgr=0.4437,box=0.0814,cls=3.5071,copy_paste=0.7000,degrees=37.3318,fliplr=0.7200,flipud=0.5856,hsv_h=0.0168,hs_2024-09-17_09-58-02/mycoco/train`

The solution is, I think, to store internally absolute paths, so we aren't affected by RayTune.

**Comment by glenn-jocher** [comment]
@Dapid thank you for your feedback. The behavior you're experiencing is due to Ray Tune's handling of working directories. To address this, you can set the environment variable `RAY_CHDIR_TO_TRIAL_DIR=0` to prevent Ray from changing directories. This should help ensure that Ultralytics-generated files are saved in the expected locations.
